### PR TITLE
Chat redux

### PR DESCRIPTION
**Pull request by TeroOtavia**
Completely reworked Muikku chat:
- Muikku chat works like "discounted" Discord

New chat features:
- drag around view
- fullscreen mode
- keep count unread messages
- block users
- ui rework
- instansed discussion

### CODE REVIEW

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/ChatController.java [hunk 1]
```diff
@@ -1,131 +1,801 @@
...
+    }
+  }
+
+  private void handleUserLeave(Long userEntityId, ChatLeaveType type) {
+    
+    // Remove all session data related to the user
+    
+    Set<String> sessionIds = userSessions.get(userEntityId);
+    if (sessionIds != null) {
+      for (String sessionId : sessionIds) {
```
Comment (aviljakaine) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/ChatController.java:618]: Could this have concurrency problems, the sessionIds set could be modified/iterated by two threads at the same time? Not a likely occurence for sure.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/ChatController.java [hunk 2]
```diff
@@ -1,131 +1,801 @@
...
+        sessionUsers.remove(sessionId);
+      }
+    }
+    userSessions.remove(userEntityId);
+    Set<Long> roomIds = roomUsers.keySet();
+    for (Long roomId : roomIds) {
+      Set<Long> users = roomUsers.get(roomId);
```
Comment (aviljakaine) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/ChatController.java:625]: This set might also have concurrency issues same as above.

**Review by aviljakaine — CHANGES_REQUESTED** [review]
Mainly the message sanitization should be added and some potential concurrency issues, otherwise just some small notes.

WebSocketMessenger is Singleton, it might warrant a check if it can become a hindrance in the future if the traffic increases. (Also the looping could be better with entryset iirc and the methods seem like they could be combined.) Not in scope for this PR though imo.

Shares db version with #7055

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/rest/ChatRESTService.java [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+    }
+    ChatMessage chatMessage = chatController.findChatMessageByIdAndArchived(id, Boolean.FALSE);
+    if (chatMessage == null) {
+      return Response.status(Status.NOT_FOUND).entity(String.format("Message %d not found", id)).build();
+    }
+    if (!sessionController.getLoggedUserEntity().getId().equals(chatMessage.getSourceUserEntityId())) {
+      return Response.status(Status.FORBIDDEN).build();
+    }
+    
+    chatController.updateMessage(chatMessage, payload.getMessage(), sessionController.getLoggedUserEntity());
```
[the remark this answers is not in the document]
Comment (aviljakaine) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/rest/ChatRESTService.java:476]: (Sanitizations could be on the controller level also.)

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/rest/ChatRESTService.java [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+      restRooms.add(restRoom);
+    }
+    return Response.ok(restRooms).build();
+  }
+  
+  @Path("/message/{IDENTIFIER}")
+  @POST
+  @RESTPermit(handling = Handling.INLINE, requireLoggedIn = true)
+  public Response postMessage(@PathParam("IDENTIFIER") String identifier, MessageContentRestModel payload) {
+    if (StringUtils.isEmpty(payload.getMessage())) {
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/rest/ChatRESTService.java:403]: The message is always handled as plain text.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/rest/ChatRESTService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/chat/rest/ChatRESTService.java:476]: Added TODO comments should messages ever be formatted in the future.

**Review by avukkonen — APPROVED** [review]

**Review by aviljakaine — APPROVED** [review]

**Pull request merged**